scipy/stats/_qmc_cy.pyx

```diff
@@ -0,0 +1,326 @@
+
+    for i in range(istart, istop):
+        for j in range(sample_view.shape[0]):
+            prod = 1
+            for k in range(sample_view.shape[1]):
+                prod *= (
+                        1 - max(sample_view[i, k], sample_view[j, k])
+                )
+            tmp_sum += prod
+
+        disc2 += tmp_sum
+        tmp_sum = 0
+
+    return disc2
+
+
+def _cy_wrapper_update_discrepancy(double[::1] x_new_view,
+                                 double[:, ::1] sample_view,
+                                 double initial_disc):
+    return c_update_discrepancy(x_new_view, sample_view, initial_disc)
+
+
+cdef double c_update_discrepancy(double[::1] x_new_view,
+                                 double[:, ::1] sample_view,
+                                 double initial_disc):
+    cdef:
+        Py_ssize_t n = sample_view.shape[0] + 1
+        Py_ssize_t xnew_nlines = x_new_view.shape[0]
+        Py_ssize_t i = 0, j = 0, k = 0
+        double prod = 1, tmp_sum= 0
+        double  disc1 = 0, disc2 = 0, disc3 = 0
+        double[::1] abs_ = np.zeros(n, dtype=np.float64)
+
+
+    # derivation from P.T. Roy (@tupui)
+    for i in range(xnew_nlines):
+        abs_[i] = fabs(x_new_view[i] - 0.5)
+        prod *= (
+                    1 + 0.5 * abs_[i]
+                    - 0.5 * pow(abs_[i], 2)
+        )
+
+    disc1 = (- 2 / <double> n) * prod
+
+    prod = 1
+    for i in range(n - 1):
+        for j in range(xnew_nlines):
+            prod *= (
+                1 + 0.5 * abs_[j] +
+                0.5 * fabs(sample_view[i, j] - 0.5) -
+                0.5 * fabs(x_new_view[j] - sample_view[i, j])
+            )
+        disc2 += prod
+        prod = 1
+
+    disc2 *= 2 / pow(n, 2)
+
+    for i in range(xnew_nlines):
+        prod *= 1 + abs_[i]
+
+    disc3 = 1 / pow(n, 2) * prod
+
+    return initial_disc + disc1 + disc2 + disc3
+
+
+ctypedef double (*func_type)(double[:, ::1], Py_ssize_t,
+                             Py_ssize_t) nogil
+
+
+cdef double threaded_loops(func_type loop_func, double[:,
+                         ::1] sample_view, unsigned int workers) nogil:
+    cdef:
+        Py_ssize_t n = sample_view.shape[0]
+        double disc2 = 0
+
+    if workers <= 1:
+        return loop_func(sample_view, 0, n)
+
+    cdef:
+        vector[thread] threads
+        unsigned int tid
+        Py_ssize_t istart, istop
+
+    for tid in range(workers):
+        istart = <Py_ssize_t> (n / workers * tid)
+        istop = <Py_ssize_t> (
+                n / workers * (tid + 1)) if tid < workers - 1 else n
+        threads.push_back(
+            thread(one_thread_loop, loop_func, ref(disc2),
+                   sample_view, istart, istop))
+
+    for tid in range(workers):
+        threads[tid].join()
+
+    return disc2
+
+
+cdef void one_thread_loop(func_type loop_func, double& disc, double[:,
+                         ::1] sample_view, Py_ssize_t istart, Py_ssize_t istop) nogil:
```

Comment:
```suggestion
cdef void one_thread_loop(func_type loop_func, double& disc, double[:,
                          ::1] sample_view, Py_ssize_t istart, Py_ssize_t istop) nogil:
```

Reply:
Same as before, insert a line break in a memory view type seems a bad idea for me.

Reply:
Sure sure! Sorry about these.